### DISCUSSION TITLE
Custom Model using KServe API is failing

### DISCUSSION THREAD

**Discussion by kjoth** [kubeflow/kfserving#3566] [Q&A]
In the inference.py when we mention the workers in the ModelServer, then the building the docker image and running it fails with below error.

if __name__ == "__main__":
    model = KServeBERTSentimentModel("onnx-all-distilroberta-v1")
    model.load()

    model_server = ModelServer**(workers=4)**
    model_server.start([model])
    
    Error
    
     docker run <prvivate-space>/onnx-roberta-model
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
INFO:kserve:Registering model: onnx-all-distilroberta-v1
INFO:kserve:Setting max asyncio worker threads as 9
INFO:kserve:Starting uvicorn with 4 workers
INFO:kserve:Starting gRPC server on [::]:8081
Traceback (most recent call last):
  File "/app/onnx_inference.py", line 87, in <module>
    model_server.start([model])
  File "/usr/local/lib/python3.11/site-packages/kserve/model_server.py", line 226, in start
    asyncio.run(servers_task())
  File "/usr/local/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 654, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/kserve/model_server.py", line 224, in servers_task
    await asyncio.gather(*servers)
  File "/usr/local/lib/python3.11/site-packages/kserve/model_server.py", line 211, in serve
    multiprocessing.set_start_method('fork')
  File "/usr/local/lib/python3.11/multiprocessing/context.py", line 247, in set_start_method
    raise RuntimeError('context has already been set')
RuntimeError: context has already been set

**Comment by sivanantha321** [comment]
Are you using the method `multiprocessing.set_start_method` somewhere else in your source code ?

**Comment by kjoth** [comment]
HI @sivanantha321 ,

The multiprocessing.set_start_method('fork') function is utilized within the Kserve model server (kserve/model_server.py). Here is the transformers_inference.py script that is employed alongside it.

`
from typing import Dict 

import kserve
import json
from transformers import AutoTokenizer, AutoModel
import torch
import torch.nn.functional as F
from kserve import ModelServer, InferRequest
import logging

def mean_pooling(model_output, attention_mask):
    token_embeddings = model_output[0] #First element of model_output contains all token embeddings
    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)

class KServeDistilrobertaModel(kserve.Model):

    def __init__(self, name: str):
        super().__init__(name)
        KSERVE_LOGGER_NAME = 'kserve'
        self.logger = logging.getLogger(KSERVE_LOGGER_NAME)
        self.logger.setLevel(logging.DEBUG)
        self.name = name
        self.ready = False

    def load(self):
        # Build tokenizer and model
        name = "sentence-transformers/all-distilroberta-v1"
        # Load model from HuggingFace Hub
        self.tokenizer = AutoTokenizer.from_pretrained(name)
        self.model = AutoModel.from_pretrained(name)
        self.ready = True


    def predict(self, request: Dict, headers: Dict) -> Dict:
        
        # self.logger.info(f"request:-- {request}")
        type(request)
        # sentences = request['input']
        sentences = request['instances'][0]['input']
        # self.logger.info(f"sequence:-- {sequence}")

        # Tokenize sentences
        encoded_input = self.tokenizer(sentences, padding=True, truncation=True, return_tensors='pt')

        # Compute token embeddings
        with torch.no_grad():
            model_output = self.model(**encoded_input)

        # Perform pooling
        sentence_embeddings = mean_pooling(model_output, encoded_input['attention_mask'])

        # Normalize embeddings
        sentence_embeddings = F.normalize(sentence_embeddings, p=2, dim=1)

        # print("Sentence embeddings: {}".format(sentence_embeddings))
        return {"predictions": sentence_embeddings.tolist()}


if __name__ == "__main__":

  model = KServeDistilrobertaModel("all-distilroberta-v1")
  model.load()

  model_server = ModelServer()
  model_server.start([model])`
  
  Running this inference.py using `python transformers_inference.py --workers 2` in local machine works perfectly.
  
<img width="691" alt="image" src="https://github.com/kserve/kserve/assets/24207765/08daff3c-3ab2-4237-9a77-ad5d185acbae">


The problem occurs after building the Docker image using the Dockerfile and attempting to include the --workers argument in the docker run command. When the --workers argument is not provided, docker run functions perfectly, but it encounters issues when the --workers argument is included.

**Without --worker args**
`docker run tf-roberta-model-args`
<img width="819" alt="image" src="https://github.com/kserve/kserve/assets/24207765/da88ea2e-dc92-4f91-bc33-47c07a6a6584">


**With --worker args**
`docker run tf-roberta-model-args --workers 2`

![image](https://github.com/kserve/kserve/assets/24207765/1ede7c97-7460-49ee-b0b1-2782a8727877)

**Comment by kjoth** [comment]
Hello @sivanantha321 ,By the way, is this problem associated with this pull request? [https://github.com/kserve/kserve/pull/3407. ](url)
